File: JavaCommon/src/client/PlayerCommand.java
package client;

import proto.ProtoPlayer.Command;
import proto.ProtoPlayer.Command.Information;
import proto.ProtoPlayer.Command.Type;

/**
 * Defines the wrapper for the proto command.
 *
 * @author cmihail (Mihail Costea)
 */
public class PlayerCommand {

  private final Type type;
  private final String info;

  /**
   * @param type the type of the command
   */
  public PlayerCommand(Type type) {
    this(type, null);
  }

  /**
   * @param type the type of the command
   * @param info information about the command
   */
  public PlayerCommand(Type type, String info) {
    this.type = type;
    this.info = info;
  }

  /**
   * @return the type of the command
   */
  public Type getType() {
    return type;
  }

  /**
   * @return the information about the command, or null if none
   */
  public String getInfo() {
    return info;
  }

  /**
   * @return the command as a proto
   */
  public Command toProto() {
    Command.Builder builder = Command.newBuilder().setType(type);
    if (info != null) {
      builder.setInfo(Information.newBuilder().setValue(info).build());
    }
    return builder.build();
  }
}

File: Server/src/proto/PlayerCommand.cpp
/*
 * PlayerCommand.cpp
 *
 *  Created on: Sep 27, 2012
 *      Author: cmihail (Mihail Costea)
 */

#include "proto/PlayerCommand.h"

#define EMPTY_STRING ""

using namespace std;

PlayerCommand::PlayerCommand(proto::Command::Type type) : type(type), info(EMPTY_STRING) {
  // Nothing to do
}

PlayerCommand::PlayerCommand(proto::Command::Type type, string info) : type(type), info(info) {
  // Nothing to do
}

PlayerCommand::PlayerCommand(Message & codedMessage) {
  proto::Command * command = new proto::Command();
  parseFromCodedMessage(codedMessage, command);

  // Set type and info.
  type = command->type();
  if (command->has_info()) {
    info = command->info().value();
  } else {
    info = EMPTY_STRING;
  }

  delete command;
}

proto::Command::Type PlayerCommand::getType() {
  return type;
}

bool PlayerCommand::hasInfo() {
  if (info.compare(EMPTY_STRING) == 0) {
    return true;
  }
  return false;
}

string PlayerCommand::getInfo() {
  return info;
}

google::protobuf::Message * PlayerCommand::toProto() {
  proto::Command * command = new proto::Command();
  command->set_type(type);
  if (info.compare(EMPTY_STRING)) {
    proto::Command::Information commandInfo = command->info();
    commandInfo.set_value(info);
  }
  return command;
}
